server/schedule/labeler/labeler.go

```diff
@@ -0,0 +1,270 @@
+		}
+	}
+	l.buildRangeList()
+	return nil
+}
+
+func (l *RegionLabeler) adjustRule(rule *LabelRule) error {
+	if rule.ID == "" {
+		return errs.ErrRegionRuleContent.FastGenByArgs("empty rule id")
+	}
+	if len(rule.Labels) == 0 {
+		return errs.ErrRegionRuleContent.FastGenByArgs("no region labels")
+	}
+	for _, l := range rule.Labels {
+		if l.Key == "" {
+			return errs.ErrRegionRuleContent.FastGenByArgs("empty region label key")
+		}
+		if l.Value == "" {
+			return errs.ErrRegionRuleContent.FastGenByArgs("empty region label value")
+		}
+	}
+
+	switch rule.RuleType {
+	case KeyRange:
+		data, ok := rule.Rule.(map[string]interface{})
+		if !ok {
+			return errs.ErrRegionRuleContent.FastGenByArgs(fmt.Sprintf("invalid rule type: %T", reflect.TypeOf(rule.Rule)))
+		}
+		startKey, ok := data["start_key"].(string)
+		if !ok {
+			return errs.ErrRegionRuleContent.FastGenByArgs(fmt.Sprintf("invalid startKey type: %T", reflect.TypeOf(data["start_key"])))
+		}
+		endKey, ok := data["end_key"].(string)
+		if !ok {
+			return errs.ErrRegionRuleContent.FastGenByArgs(fmt.Sprintf("invalid endKey type: %T", reflect.TypeOf(data["end_key"])))
+		}
+		var r KeyRangeRule
+		r.StartKeyHex, r.EndKeyHex = startKey, endKey
+		var err error
+		r.StartKey, err = hex.DecodeString(r.StartKeyHex)
+		if err != nil {
+			return errs.ErrHexDecodingString.FastGenByArgs(r.StartKeyHex)
+		}
+		r.EndKey, err = hex.DecodeString(r.EndKeyHex)
+		if err != nil {
+			return errs.ErrHexDecodingString.FastGenByArgs(r.EndKeyHex)
+		}
+		if len(r.EndKey) > 0 && bytes.Compare(r.EndKey, r.StartKey) <= 0 {
+			return errs.ErrRegionRuleContent.FastGenByArgs("endKey should be greater than startKey")
+		}
+		rule.Rule = &r
+		return nil
+	}
+	log.Error("invalid rule type", zap.String("rule-type", rule.RuleType))
+	return errs.ErrRegionRuleContent.FastGenByArgs(fmt.Sprintf("invalid rule type: %s", rule.RuleType))
+}
+
+func (l *RegionLabeler) buildRangeList() {
+	builder := rangelist.NewBuilder()
+	for _, rule := range l.labelRules {
+		if rule.RuleType == KeyRange {
+			r := rule.Rule.(*KeyRangeRule)
+			builder.AddItem(r.StartKey, r.EndKey, rule)
+		}
+	}
+	l.rangeList = builder.Build()
+}
+
+// GetAllLabelRules returns all the rules.
+func (l *RegionLabeler) GetAllLabelRules() []*LabelRule {
+	l.RLock()
+	defer l.RUnlock()
+	rules := make([]*LabelRule, 0, len(l.labelRules))
+	for _, rule := range l.labelRules {
+		rules = append(rules, rule)
+	}
+	return rules
+}
+
+// GetLabelRules returns the rules that match the given ids.
+func (l *RegionLabeler) GetLabelRules(ids []string) ([]*LabelRule, error) {
+	l.RLock()
+	defer l.RUnlock()
+	rules := make([]*LabelRule, 0, len(ids))
+	for _, id := range ids {
+		if rule, ok := l.labelRules[id]; ok {
+			rules = append(rules, rule)
+		} else {
+			return nil, errs.ErrRegionRuleNotFound.FastGenByArgs(id)
+		}
+	}
+	return rules, nil
+}
+
+// GetLabelRule returns the Rule with the same ID.
+func (l *RegionLabeler) GetLabelRule(id string) *LabelRule {
```

Comment:
For now, the implementation of TiDB hasn't used `GetLabelRule` yet. Perhaps, it can be replaced by `GetLabelRules` with only one ID in the list.